### ISSUE TITLE
Get DbContext instance from EntityFrameworkManager.ContextFactory

### ISSUE THREAD

**Issue by MaklaCof** [zzzprojects/EntityFramework-Plus#217]
First: do I need to set `EntityFrameworkManager.ContextFactory` in ConfigureServices in .NET Core 2.0?
And second: how can I get instance of DbContext from ContextFactory? 

I need new DbContext instance in Task and I want to use your factory, but don't know how can I get it?

**Comment by JonathanMagnan** [comment]
Hello @MaklaCof ,

I'm not sure about this question since our other library (Entity Framework Extensions) also have an `EntityFrameworkManager.ContextFactory` property.

From which library do you speak?

Best Regards,

Jonathan

**Comment by MaklaCof** [comment]
```
namespace Z.EntityFramework.Extensions
{
    public class EntityFrameworkManager
    {
        public static Func<DbContext, DbContext> ContextFactory;

        public EntityFrameworkManager();
    }
}
```
My bad. I was upgrading project to 2.0 and run at this code in startup.

**Comment by JonathanMagnan** [comment]
Hello @MaklaCof ,

Setting this function is required to make EF Core work if your context doesn't have a default constructor.

I normally recommend doing it in a static constructor in your context

```csharp
static MyContextClass()
{
	var options = new DbContextOptionsBuilder();
	options.UseSqlServer(new SqlConnection(My.Config.ConnectionStrings.TestDatabase));
	EntityFrameworkManager.ContextFactory = context => new MyContextClass(options);
}
```

This member `ContextFactory` doesn't provide you the current context but a new context which is normally used by our code to retrieve additional information without affecting the current context.

Best Regards,

Jonathan

**Comment by MaklaCof** [comment]
I set
`EntityFrameworkManager.ContextFactory = context => new MyContextClass(options);`
in Startup.cs -> ConfigureServices. Is this not the right place?

And second: how can I get Context instance from your factory? Same thing as you do in your library when you need a new Context?

**Comment by JonathanMagnan** [comment]
Hello @MaklaCof ,

It's a good place! Any place is good as long as the initialize is done only once.

From our library, we normally already know the `true` context since it's passed in our method.

By example:

```csharp
public static void BulkInsert<T>(this DbContext @this, IEnumerable<T> entities, Action<EntityBulkOperation<T>> bulkOperationFactory) where T : class
{
	// ...code...

	var hookContext = Z.EntityFramework.Extensions.EntityFrameworkManager.ContextFactory(@this);
	
	// ...code...
}
```

Could you provide me an example of something you would like to do? Perhaps I could help more since I'm not exactly sure what you are looking for.

Best Regards,

Jonathan

**Comment by MaklaCof** [comment]
First thank you very much for helping me.

Something like that:
```
        private async bool MyMethod()
        {
            //some code

            Task task = new Task(() =>
            {
                var context = Z.EntityFramework.Extensions.EntityFrameworkManager.ContextFactory.GetNewInstance();
                //instead !!!!
                /******  I do exactly this in Startup when assigning  EntityFrameworkManager.ContextFactory
                var builder = new DbContextOptionsBuilder<AppDbContext>();
                builder.UseSqlServer(Configuration["ConnectionStrings:Web"]);
                if (this.env.IsDevelopment())
                    builder.EnableSensitiveDataLogging();

                var lf = new LoggerFactory();
                lf.AddProvider(new OpPISLoggerProvider(Configuration.GetSection("Log")));
                builder.UseLoggerFactory(lf);

                var context = AppDbContext(builder.Options);
                */

                context.Users.Where (....)

            });

            this.Context.Where(....);
            //more work to do

            task.Wait();
            return true;
        }
```

**Comment by JonathanMagnan** [comment]
Hello @MaklaCof ,

I'm not sure to understand why you need our library to create an instance of your context.

On our side, we don't have access to your context, so it makes senses!

For your side, you probably have access to the constructor, so why using our library?

Anyway, depending on how you have set the ContextFactory, you may not need the current context.

```csharp
Task task = new Task(() =>
{
	var context = Z.EntityFramework.Extensions.EntityFrameworkManager.ContextFactory(null);
	// ...code...
}
```

Hope it helps,

Best Regards,

Jonathan

**Comment by MaklaCof** [comment]
Because constructor with correct options takes several lines of codes and yours only one.
